subprojects/model-core/src/main/java/org/gradle/internal/reflect/annotations/impl/DefaultTypeAnnotationMetadataStore.java

```diff
@@ -349,6 +361,18 @@ private void processMethodAnnotations(Method method, Map<String, PropertyAnnotat
         }
     }
 
+    private static void validateSetterForMutableType(Method method, ValidationErrorsBuilder errorsBuilder, PropertyAccessorType accessorType) {
+        Class<?> setterType = getSetterType(method);
+        if (MUTABLE_NON_FINAL_TYPES.contains(setterType)) {
+            errorsBuilder.recordError(String.format("Support for property setters for type '%s' will be removed in a future release. Please consider making '%s' final.", setterType.getName(), accessorType.propertyNameFor(method)));
```

Comment:
`final` only makes sense in the context of Groovy. What we should be complaining about is the fact that the property has a setter, and that it should be removed.

Reply:
Also, this is not a deprecation warning. We are not planning to remove support for anything yet. We should only say that this kind of pattern shouldn't be used. Take a look at how we handle problematic cases like both field and getter being annotated.

Reply:
I've rewritten it, let me know if this is the direction you want to go toward.